soroban-env-host/src/auth.rs

```diff
@@ -928,47 +916,54 @@ impl AuthorizationTracker {
     }
 
     fn verify_nonce(&mut self, host: &Host) -> Result<(), HostError> {
-        let nonce_is_correct = if let Some(nonce) = self.maybe_consume_nonce(host)? {
-            if let Some(tracker_nonce) = self.nonce {
-                tracker_nonce == nonce
-            } else {
-                // If the nonce isn't set in the tracker, but is required, then
-                // it's incorrect.
-                false
+        if !self.need_nonce {
+            return Ok(());
+        }
+        self.need_nonce = false;
+        if let Some((nonce, expiration_ledger)) = &self.nonce {
+            if host.with_ledger_info(|li| Ok(li.sequence_number))? > *expiration_ledger {
+                return Err(host.err(
+                    ScErrorType::Auth,
+                    ScErrorCode::InvalidInput,
+                    "signature has expired",
+                    &[],
+                ));
+            }
+            if let Some(addr) = &self.address {
+                return host.consume_nonce(
+                    addr.metered_clone(host.budget_ref())?,
+                    *nonce,
+                    *expiration_ledger,
+                );
             }
-        } else {
-            // Nonce is either already checked or not needed in the first place.
-            true
-        };
-        if nonce_is_correct {
-            Ok(())
-        } else {
-            Err(host.err(
-                ScErrorType::Auth,
-                ScErrorCode::InvalidInput,
-                "nonce is incorrect",
-                &[],
-            ))
         }
+        Err(host.err(
+            ScErrorType::Auth,
+            ScErrorCode::InternalError,
+            "unexpected nonce verification state",
+            &[],
+        ))
     }
 
     // Computes the payload that has to be signed in order to authenticate
     // the authorized invocation tree corresponding to this tracker.
     fn get_signature_payload(&self, host: &Host) -> Result<[u8; 32], HostError> {
+        let (nonce, expiration_ledger) = self.nonce.ok_or_else(|| {
+            host.err(
+                ScErrorType::Auth,
+                ScErrorCode::InternalError,
+                "unexpected missing nonce",
+                &[],
+            )
+        })?;
         let payload_preimage =
             HashIdPreimage::SorobanAuthorization(HashIdPreimageSorobanAuthorization {
                 network_id: Hash(
                     host.with_ledger_info(|li| li.network_id.metered_clone(host.budget_ref()))?,
                 ),
+                nonce,
+                signature_expiration_ledger: expiration_ledger,
```

Comment:
If someone bumps the expiration ledger of the nonce, will the signature become invalid?

Reply:
If nonce entry is bumped, then this nonce won't be usable for the owner's address. That's fine as long as someone doesn't try to always use the same nonce (which kind of defeats its purpose).

The signature is invalidated purely based off of the current ledger.